src/util/object_enum.rs

```diff
@@ -0,0 +1,107 @@
+//! Helper types for object enumeration
+
+use std::marker::PhantomData;
+
+use crate::vm::VMBinding;
+
+use super::{
+    heap::{
+        chunk_map::{ChunkMap, ChunkState},
+        MonotonePageResource,
+    },
+    linear_scan::Region,
+    metadata::side_metadata::spec_defs::VO_BIT,
+    Address, ObjectReference,
+};
+
+/// A trait for enumerating objects in spaces, used by [`Space::enumerate_objects_coarse`].
+///
+/// [`Space::enumerate_objects_coarse`]: crate::policy::space::Space::enumerate_objects_coarse
+pub trait ObjectEnumerator {
+    /// Visit a single object.
+    fn visit_object(&mut self, object: ObjectReference);
+    /// Visit an address range that may contain objects.
+    fn visit_address_range(&mut self, start: Address, end: Address);
+}
+
+/// An implementation of `ObjectEnumerator` that wraps a callback.
+pub struct ClosureObjectEnumerator<F, VM>
+where
+    F: FnMut(ObjectReference),
+    VM: VMBinding,
+{
+    object_callback: F,
+    phantom_data: PhantomData<VM>,
+}
+
+impl<F, VM> ClosureObjectEnumerator<F, VM>
+where
+    F: FnMut(ObjectReference),
+    VM: VMBinding,
+{
+    pub fn new(object_callback: F) -> Self {
+        Self {
+            object_callback,
+            phantom_data: PhantomData,
+        }
+    }
+}
+
+impl<F, VM> ObjectEnumerator for ClosureObjectEnumerator<F, VM>
+where
+    F: FnMut(ObjectReference),
+    VM: VMBinding,
+{
+    fn visit_object(&mut self, object: ObjectReference) {
+        (self.object_callback)(object);
+    }
+
+    fn visit_address_range(&mut self, start: Address, end: Address) {
+        VO_BIT.scan_non_zero_values::<u8>(start, end, &mut |address| {
+            let object = ObjectReference::from_address::<VM>(address);
+            (self.object_callback)(object);
+        })
+    }
+}
+
+/// Allow querying if a block may have objects. `MarkSweepSpace` and `ImmixSpace` use different
+/// `Block` types, and they have different block states. This trait lets both `Block` types provide
+/// the same `may_have_objects` method.
+pub(crate) trait BlockMayHaveObjects: Region {
```

Comment:
Why not add the method to `Region`?

Reply:
We only need it for the two `Block` types, one for `MarkSweepSpace` and the other for `ImmixSpace`.  Adding the method to `Region` will require all types that implement `Region` to implement that method, which is not good because

1.  It is unnecessary because we currently don't do the same testing on other levels, such as `Line`.  For `Chunk`, all spaces use the same `Chunk` type, and it is unnecessary to make it implement a generic trait.
2.  `Region` is a kind of address range (power of 2 sized and aligned), and all its methods are about address computing and comparing.  But the method `may_have_objects` is about querying its metadata.
3.  Not all `Region` types have enough metadata to identify if it may contain any object.  `Line`, for example, only has the line marking bits after tracing, and is unhelpful for traversing the heap during mutator time.